libraries/common/stream.effekt

```diff
@@ -0,0 +1,417 @@
+    replicate(number - 1) {action}
+  }
+
+
+def sum[R] { stream : () => R / emit[Int] }: (R, Int) = {
+  var s = 0;
+  try {
+    (stream(), s)
+  } with emit[Int] { v =>
+    s = s + v;
+    resume(())
+  }
+}
+
+def sum { stream : () => Unit / emit[Int] }: Int =
+  sum[Unit]{stream}.second
+
+def collectList[A, R] { stream: () => R / emit[A] }: (R, List[A]) =
+  try {
+    (stream(), Nil())
+  } with emit[A] {
+    def emit(v) = {
+      val (r, vs) = resume(());
+      (r, Cons(v, vs))
+    }
+  }
+
+def collectList[A] { stream: () => Unit / emit[A] }: List[A] =
+  collectList[A, Unit]{stream}.second
+
+def collectArray[A, R] { stream: () => R / emit[A] }: (R, Array[A]) = {
+  var i = 0
+  var a = array::allocate(1)
+  try {
+    (stream(), a.resize(i))
+  } with emit[A] {
+    def emit(v) = {
+      if (i >= a.size) { a = a.resize(2 * a.size) }
+      a.unsafeSet(i, v)
+      i = i + 1
+      resume(())
+    }
+  }
+}
+
+def collectArray[A] { stream: () => Unit / emit[A] }: Array[A] =
+  collectArray[A, Unit]{stream}.second
+
+def collectBytes[R] { stream: () => R / emit[Byte] }: (R, ByteArray) = {
+  var i = 0
+  var a = bytearray::allocate(1)
+  try {
+    (stream(), a.resize(i))
+  } with emit[Byte] {
+    def emit(v) = {
+      if (i >= a.size) { a = a.resize(2 * a.size) }
+      a.unsafeSet(i, v)
+      i = i + 1
+      resume(())
+    }
+  }
+}
+
+def collectBytes { stream: () => Unit / emit[Byte] }: ByteArray =
+  collectBytes[Unit]{stream}.second
+
+
+effect read[A](): A / stop
+
+
+def feed[T](list: List[T]) { reader: () => Unit / read[T] } = {
+  var l = list
+  try {
+    reader()
+  } with read[T] { () =>
+    resume {
+      l match {
+        case Nil() => do stop()
+        case Cons(value, rest) =>
+          l = rest
+          return value
+      }
+    }
+  }
+}
+
+def feed[T](array: Array[T]) { reader: () => Unit / read[T] } = {
+  var i = 0
+  try {
+    reader()
+  } with read[T] { () =>
+    resume {
+      if (i < array.size) {
+        array.unsafeGet(i)
+      } else {
+        do stop()
+      }
+    }
+  }
+}
+
+def feed(bytes: ByteArray) { reader: () => Unit / read[Byte] } = {
+  var i = 0
+  try {
+    reader()
+  } with read[Byte] { () =>
+    resume {
+      if (i < bytes.size) {
+        bytes.unsafeGet(i)
+      } else {
+        do stop()
+      }
+    }
+  }
+}
+
+def source[A, R] { stream: () => Unit / emit[A] } { reader: () => R / read[A] }: R = {
+  var next = box { () => None() }
+  next = box { () =>
+    try {
+      stream()
+      None()
+    } with emit[A] {
+      def emit(v) = {
+        next = box { () => resume(()) }
+        Some(v)
+      }
+    }
+  }
+  try {
+    reader()
+  } with read[A] { () =>
+    resume {
+      next().getOrElse { do stop() }
+    }
+  }
+}
+
+
+def writeFile[R](path: String) { stream: () => R / emit[Byte] }: R / Exception[IOError] = {
+
+  val file = open(path, WriteOnly());
+  with on[IOError].finalize { close(file) }
+
+  val chunkSize = 1048576 // 1MB
```

Comment:
It would be cool to have a separate standard library that helps with file sizes and other units, like
```
1.MB
100.ms
etc.
```